### DISCUSSION TITLE
HA tradeoffs, component location, encryption, maintenance, and sharing

### DISCUSSION THREAD

**Discussion by terribleplan** [seaweedfs/seaweedfs#3322] [General]
First, let me say that this project is awesome. I am super excited about how easy it makes distributed replicated storage. I am working toward building a cluster using many small nodes (ARM, 2 disks, gigabit networking) to replace my single-server ZFS storage, and have several questions. Any assistance whether answering the questions I pose or general commentary about my plans would be greatly appreciated.

# HA Tradeoffs

I do not know what the failure rate of the nodes themselves will be vs the failure rate of the disks, though I suspect some of the disks will fail first as some of them are a bit older from when I replaced disks in my ZFS server with higher capacity ones. I understand that the volume server can be provided with multiple disks to create volumes on, but I have also seen it stated that it can be better if a separate volume server is run for each disk.

I see a few possibilities in how I can set up the volume servers:
1. Run one volume server per node, providing it with both disks. This seems like an OK option with few disks per node. I don't know the implications/feasibility of starting the node after it has lost half of its data, is there special data that is stored only on one of the disks?
2. Run one volume server per node with the disks in RAID-0. This seems like the worst of all worlds as the failure of a single disk means a complete loss of data, and the performance benefit of software RAID-0 is probably not going to matter much compared to the limitations of gigabit networking.
3. Run one volume server per node with the disks in RAID-1. This reduces the storage capacity by a pretty decent amount since I would still want to replicate at _least_ once to be able to easily tolerate node failures or RAID failures, meaning I would always be opted in to storing 4 copies.
4. Run one volume server per disk. This seams like it would tolerate disk failures the best, but may not be optimal for node failures. Would a replication of 001 mean that sometimes _both_ copies of the volume will be on the same physical node? Would I need to make a rack per server and use a replication of 010 to solve this? 

What setup would you recommend, or what rules of thumb do you use when provisioning clusters? Is there some sort of built-in protection preventing storing replica data on the same _physical_ server even if it is running multiple logical volume servers?

# Component Location

Where should I run certain components? As a proof of concept I was running everything on the nodes themselves. This worked, but the performance wasn't great (probably due to the networking setup for the PoC) at about 22MB/s for 3 nodes, and it meant storing the data for master, volumes, and filer on the same disks/nodes.

Since the master server is stateless/easily rebuilt, doesn't seem to store much data, and shouldn't be dealing with any volume data it seems like it would be fine to run it on the same nodes as the volume servers, and there shouldn't be significant performance implications.

The filer seems like the opposite. I have already noticed slowness when doing simple operations like `fs.du`, which takes about 25 seconds for my PoC filesystem storing ~4T across 500 directories and 10k files. It seems like it would benefit from storing its data on an SSD if I continue with leveldb. Locating filer closer to the servers accessing the filesystem (or WebDAV, or S3) seems like it would be a win for latency. I am thinking it would also be better (or at least make me feel better) to store the data in something like cockroach or redis cluster where I have a good idea of how to back it up for DR and can spin up another/down filers without a care in the world.

For a real setup it seems like running it on a server with high throughput networking would be very beneficial for write performance.

# Encryption

When enabling filer encryption and using FUSE mounts does the filer hand the encryption keys to the clients to allow them to read and write directly to the filer, or do operations end up proxying through the filer?

# Maintenance

For the most part my data set is written once, infrequently accessed, and occasionally some of it will be deleted. Generally it seems like erasure coding will make sense for this dataset (and I really like the availability, reliability, and space saving of EC), but I am concerned about the deletions. How does the interaction between filer and EC volumes work when filer wants to delete something from a volume that is read-only (which is inherent to EC volumes)? Is there something that could be done in a the maintenance script to decode EC volumes to allow for deletes and vacuum to happen if there is enough of a reason to do so? Does EC not actually make sense for my use case?

# Sharing

It's neat that seaweedfs supports WebDAV, but that option will be unusable for me since the Windows WebClient (what it uses to mount/access WebDAV) is only able to read/write files <4GB (which involves editing the registry to even get to the limit), and may of my files will be above this size. It seems like it would make the most sense to run SAMBA (or similar) on some server that mounts the filesystem. Would running SAMBA, mount, and filer all on the same machine (assuming it has plenty of network throughput) would be my best choice for that?

**Comment by chrislusf** [comment]
For HA:
1. No special data for any disk or directory.
2. RAID is agnostic to SeaweedFS and is not recommended.
3. Ditto. 
4. Yes, create racks and use 010.

If using one volume server with multiple directories/disks, the same data would not be on the same server.

**Reply by terribleplan** [reply]
I think I'll probably go with multiple volume servers per host and use racks to provide the redundancy guarantees I want, thanks for the insights.

**Comment by chrislusf** [comment]
For Component Location

Correct. Filer is doing most of the work. You may want to add multiple filers for high throughput.

**Comment by chrislusf** [comment]
For Encryption:

FUSE mount will handle the encryption/decryption.

**Comment by chrislusf** [comment]
For Maintenance:

Deletion on EC volumes is only soft deletion. To really purge the garbage, the EC volumes needs to be convert to normal volumes and encode back to EC again.

**Reply by terribleplan** [reply]
Thanks for the quick replies! Is there any command to easily convert EC volumes that have crossed some threshold of soft deleted bytes/values so that the normal vacuum can take place, or will I need to inspect EC volumes for deleted bytes and call ec.decode and then wait for the vacuum to happen?

**Comment by chrislusf** [comment]
For Sharing:

SAMBA over mount may be good.

Filer and mount does not need to be on the same server for performance purpose.

**Reply by terribleplan** [reply]
For sure, I'll just want to keep in mind the tradeoff of having the metadata on the same machine vs a network hop away for the mount. I'll probably start with a single filer (also providing S3), a FUSE mount of that filer, and SAMBA all on one machine (with fast disks and multiple 10g connections) then scale/split it from there as necessary.
